### DISCUSSION TITLE
Problems with divisionals

### DISCUSSION THREAD

**Discussion by mveccia** [GrandOrgue/grandorgue#1349] [Help]
In my code I think I just wrote the various lines of a divisional, which I report below, but when I start the pipe organ these non-blocking messages appear. By the way, the divisionals work perfectly!!! 

Can anyone explain to me why? 

[Divisional001]
Name=1
;divisionale per manuale Grandorgano
NumberOfStops=0
NumberOfSwitches=3
Switch006=-006
Switch007=-007
Switch008=-008

NumberOfCouplers=0
NumberOfTremulants=0

Displayed=Y
ImageOn=Dati\Grandorgano\pistonOn.png
ImageOff=Dati\Grandorgano\pistonOff.png
PositionX=960
PositionY=575

(traslate from italian: (Invalid entry combination )

Error: Combinazione di voce non valida Switch001 in Divisional001
Error: Combinazione di voce non valida Switch002 in Divisional001
Error: Combinazione di voce non valida Switch003 in Divisional001

…

(traslate from italian: (unused ODF entry)
Warning: Voce ODF inutilizzata 'Divisional001/Switch006'
Warning: Voce ODF inutilizzata 'Divisional001/Switch007'
Warning: Voce ODF inutilizzata 'Divisional001/Switch008'

**Comment by eturpault** [comment]
I think that these warning are caused by the fact that the switches references are not starting from 1, it should be written : 

> NumberOfSwitches=3
> Switch**001**=-006
> Switch**002**=-007
> Switch**003**=-008
> 

instead of : 

> NumberOfSwitches=3
> Switch006=-006
> Switch007=-007
> Switch008=-008

**Reply by mveccia** [reply]
Thanks for the reply, but I had already tried, and it's worse: the blocking message appears
"Error: Value out of range in section 'Divisional001' item 'Switch002': -7"
What is strange is that it does it on the second Switch and not on the first!!!

[Divisional001]
Name=1
;divisionale per manuale Grandorgano
NumberOfSwitches=4
Switch001=-006
Switch002=-007
Switch003=-008
Switch004=-016

NumberOfStops=0
NumberOfCouplers=0
NumberOfTremulants=0
Displayed=Y
ImageOn=Dati\Grandorgano\pistonOn.png
ImageOff=Dati\Grandorgano\pistonOff.png
PositionX=960
PositionY=575

**Comment by eturpault** [comment]
I suppose that somewhere else in your ODF you have [Switch006], [Switch007], [Switch008] and [Switch016] sections duly defined, isn't it ?
In the [Organ] section the field NumberOfSwitches should be set at a value higher or equal to 16, is it the case ?
If the error occurs on the second Switch and not on the first, I guess that you have NumberOfSwitches=6 in the Organ section.

**Reply by mveccia** [reply]
There are 16 switches in the odf, some of which must not affect the divisionals
[Manual001]
Name=Grandorgano
.....
NumberOfSwitches=6
Switch001=004
Switch002=005
Switch003=006
Switch004=007
Switch005=008
Switch006=016

[Organ]
.....
NumberOfSwitches=16

**Reply by larspalo** [reply]
And have you tried to specify in the [Divisional999]
Switch001=-001
Switch002=-002
etc?
Since you have already have defined which switches should be available for the manual that should be enough and you just specify with the - or positive number if it's on or off.

By the way, nowadays it's highly recommended to not use the divisionals from the ODF specification as the general approach. Only if there's a very specific reason to use them - like specifying fixed combinations, or some kind of "built in presets" they really fill a purpose. The built in divisionals from GO (that one also find in the divisional panel) that you just specify as GUIElement will work just fine instead and requires much fewer lines of ODF code with no disadvantages to the user.

**Reply by mveccia** [reply]
Dear Lars, thank you for your reply. I mainly need the divisionails because I care a lot about the graphic construction of a real organ, and if that organ has divisional buttons it is my necessity to present them "working" in the console. However I must say that the original ODF code, namely:
[Divisional001]
Name=1
;divisional for manual Grandorgano
NumberOfStops=0
NumberOfSwitches=3
Switch006=-006
Switch007=-007
Switch008=-008
it is non-blocking and the buttons are perfectly programmable. Only the annoyance of warnings that can create disorientation remains. If you have any more tips, I'd be delighted.
![immagine](https://user-images.githubusercontent.com/46837801/214148048-f0b814a7-9cd9-4343-97c4-cf0be9a955bf.png)

**Reply by larspalo** [reply]
Ok, I attach an organ file that display both methods to include divisionals. For the pedal I used the old style specified in ODF and for manual 1 and 2 I used the newer setter element style. Both will work for "general" usage. The divisionals specified in ODF are best used for special cases as mentioned before while the setter type divisionals is much easier to include (and is more capable...).

[DivisionalsTest.zip](https://github.com/GrandOrgue/grandorgue/files/10489954/DivisionalsTest.zip)

**Comment by willeke1234** [comment]
In divisional line `Switch002=-007` the number 002 on the left side is the index of the switch in the divisional. The number 007 on the right side is the index of the switch in the division/manual. When you have a manual

```
[Manual001]
Name=Grandorgano
.....
NumberOfSwitches=6
Switch001=004
Switch002=005
Switch003=006
Switch004=007
Switch005=008
Switch006=016
```

and want switches 006, 007, 008 and 016 in the divisional then the divisional is

```
[Divisional001]
Name=1
;divisionale per manuale Grandorgano
NumberOfSwitches=4
Switch001=-003
Switch002=-004
Switch003=-005
Switch004=-006
```

**Reply by larspalo** [reply]
This is correct, but I might add that the only reason to add the Switches to the [Manual999] section to begin with is to have them included in the divisionals so it mostly makes sense to include them all in the [Divisional999] section too.

Anyway, my previous statement about the divisional usage still apply. Use [Divisional999] sections if you really need to specify for instance fixed combinations or presets for the division that the organ you model has. If just a programmable divisional for the user is required - use instead a [Panel999Element999] with Type=SetterXXXDivisionalYYY. It's fewer lines of ODF code and lets GO take care of the internals, and you have access to more divisionals than through the ODF.

**Reply by mveccia** [reply]
it is impossible: if there are more than one manuals, the switch with the same number is indicated several times... in the group [divisionalxxx] there is no reference to which manual it refers to:

[Manual000]
Name=Pedaliera
.....
NumberOfSwitches=5
Switch001=001
Switch002=002
Switch003=003
Switch004=014
Switch005=015

[Manual001]
Name=Grandorgano
.....
NumberOfSwitches=6
Switch001=004
Switch002=005
Switch003=006
Switch004=007
Switch005=008
Switch006=016

[Manual002]
Name=Positivo
.....
NumberOfSwitches=5
Switch001=009
Switch002=010
Switch003=011
Switch004=012
Switch005=013

I'll try the Lars method

**Reply by larspalo** [reply]
> it is impossible: if there are more than one manuals, the switch with the same number is indicated several times... in the group [divisionalxxx] there is no reference to which manual it refers to:

The divisional itself belongs to one specific manual, so the referencing is not a problem at all. But both the methods I show in the example ODF (which can be loaded in GO) can be used and I suggest you use the

[Panel999Element999]
Type=SetterXXXDivisionalYYY

method if you don't have a _very_ good reason not to.

Also, only specify in the [Manual999] section the switches you actually want to have stored in the divisionals, not all semi related ones.
